gateway_code/nodes.py

```diff
@@ -0,0 +1,184 @@
+# -*- coding:utf-8 -*-
+
+# This file is a part of IoT-LAB gateway_code
+# Copyright (C) 2015 INRIA (Contact: admin@iot-lab.info)
+# Contributor(s) : see AUTHORS file
+#
+# This software is governed by the CeCILL license under French law
+# and abiding by the rules of distribution of free software.  You can  use,
+# modify and/ or redistribute the software under the terms of the CeCILL
+# license as circulated by CEA, CNRS and INRIA at the following URL
+# http://www.cecill.info.
+#
+# As a counterpart to the access to the source code and  rights to copy,
+# modify and redistribute granted by the license, users are provided only
+# with a limited warranty  and the software's author,  the holder of the
+# economic rights,  and the successive licensors  have only  limited
+# liability.
+#
+# The fact that you are presently reading this means that you have had
+# knowledge of the CeCILL license and that you accept its terms.
+
+""" Common logic for plugin nodes classes """
+import abc
+import os
+import pkgutil
+
+from gateway_code.utils import elftarget
+
+
+class MetaNodeBase(type):
+    """ Base Metaclass for Plugin Control or Open Nodes"""
+
+    def __init__(cls, name, bases, class_dict):
+        type.__init__(cls, name, bases, class_dict)
+        if not hasattr(cls, '__registry__'):
+            cls.__registry__ = {}
+        else:
+            cls.__registry__[cls.TYPE] = cls
+
+
+class MetaNode(MetaNodeBase, abc.ABCMeta):
+    """ Combining metaclasses,
+    control or open Nodes have a contract verified on instantiation """
+    pass
+
+
+class NodeBase(object):  # pylint: disable = too-few-public-methods
```

Comment:
I don't understand why you need 3 extra levels of abstract class here. Why not just do this:
```python
class NodeBase(metaclass=ABCMeta):

    def __init__(self):
        type.__init__(cls, name, bases, class_dict)
        if not hasattr(cls, '__registry__'):
            cls.__registry__ = {}
        else:
            cls.__registry__[cls.TYPE] = cls

    @abc.abstractmethod
    def status(self):
        pass

class ControlNode(NodeBase):
    TYPE = 'undefined'

# or maybe ? (not sure about that)

class ControlNode(metaclass=NodeBase):
    TYPE = 'undefined'
```

Reply:
It's because the Node metaclass need to inherit from both ABCMeta and the 'registry' metaclass, to have both behaviors. MetaNode is a combining metaclass, which is the metaclass for the ControlNode and OpenNode classes. 

Essentially, when seeing `class MyNode(OpenNode)`, Python looks for a metaclass in OpenNode, finds that it's MetaNode, and adds *both* a `__registry__` attribute to the MyNode class, and the way to check that the MyNode subclass implements the abstract methods defined in OpenNode.

i'll add some unit tests to demonstrate that

We could remove the NodeBase at the cost of a repeat of the `status` method though